src/ranger_var.c

```diff
@@ -0,0 +1,156 @@
+#include <R.h>
+#include <Rinternals.h>
+#include <math.h>
+
+/* 
+//FIXME:
+    * make most funs static
+*/
+
+// Debug printer system - can be switched on/off
+#define DEBUG_ENABLED 1  // Set to 1 to enable debug output
+
+#if DEBUG_ENABLED
+#define DEBUG_PRINT(fmt, ...) Rprintf(fmt, ##__VA_ARGS__)
+#else
+#define DEBUG_PRINT(fmt, ...) do {} while(0)
+#endif
+
+
+// for tree with index j_tree:
+// computes a matrix with 2 columns: mu and sigma^2 for each term-node, term-nodes are rows
+// the matrix is stored in s_res, which is a list of matrices, one for each tree
+void c_ranger_mu_sigma_per_tree(int j_tree, int n_obs, int n_trees, int *pred_tab, double *y, SEXP s_res) {
+    DEBUG_PRINT("tree: %d\n", j_tree);
+    int n_terms = 0; // FIXME: careful some ids might be missing
+    int term; 
+    // get max term-node id, we need to know the size of the arrays
+    for (int i = 0; i < n_obs; i++) {
+      term = pred_tab[i + j_tree * n_obs];
+      if (term > n_terms) n_terms = term;
+    }
+    DEBUG_PRINT("n_terms: %d\n", n_terms);
+    // +1 because our C arrays are 0-indexed
+    int counts[n_terms + 1]; // 
+    double means[n_terms + 1]; // 
+    double m2s[n_terms + 1]; //
+    memset(counts, 0, sizeof(counts));
+    memset(means, 0, sizeof(means));
+    memset(m2s, 0, sizeof(m2s));
+
+    // Welford's variance algorithm (numerically stable)
+    for (int i = 0; i < n_obs; i++) {
+      term = pred_tab[i + j_tree * n_obs];
+      counts[term]++;
+      double yy = y[i];
+      double delta = yy - means[term];
+      means[term] += delta / counts[term];
+      m2s[term] += delta * (yy - means[term]);
+      DEBUG_PRINT("i: %d, term: %d, yy: %f, delta: %f, means: %f, m2s: %f\n", i, term, yy, delta, means[term], m2s[term]);
+    }
+
+    // Create result array with dimensions [n_terms, 2] for mu and sigma^2
+    SEXP s_mu_sigma2_mat = PROTECT(allocMatrix(REALSXP, n_terms, 2));
+    SEXP s_dimnames = PROTECT(allocVector(VECSXP, 2));
+    SEXP s_colnames = PROTECT(allocVector(STRSXP, 2));
+    SET_STRING_ELT(s_colnames, 0, mkChar("mu"));
+    SET_STRING_ELT(s_colnames, 1, mkChar("sigma2")); 
+    SET_VECTOR_ELT(s_dimnames, 1, s_colnames);
+    setAttrib(s_mu_sigma2_mat, R_DimNamesSymbol, s_dimnames);
+    
+    double *res = REAL(s_mu_sigma2_mat);
+    for (int i = 1; i <= n_terms; i++) {
+      res[i-1]           = means[i];                    
+      res[i-1 + n_terms] = m2s[i] / (counts[i] - 1);
+    }
```

Comment:
After this there will be a bunch of spurious zeros for mus and sigma2s -- does this not cause problems later on?